stan/math/prim/mat/fun/ocl_gpu.hpp

```diff
@@ -0,0 +1,292 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_OCL_HPP
+#define STAN_MATH_PRIM_MAT_FUN_OCL_HPP
+
+#define __CL_ENABLE_EXCEPTIONS
+
+#include <stan/math/prim/arr/err/check_opencl.hpp>
+#include <CL/cl.hpp>
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#define DEVICE_FILTER CL_DEVICE_TYPE_GPU
+
+/*
+ *    @file stan/math/prim/mat/fun/ocl.hpp
+ *    @brief Initialization for OpenCL: find platforms, devices,
+ *      create context, command queue etc.
+ */
+
+namespace stan {
+namespace math {
+
+inline void compile_kernel_group(std::string group);
+
+static int initialized_ = 0;
+typedef std::map<std::string, std::string> map_string;
+typedef std::map<std::string, cl::Kernel> map_kernel;
+typedef std::map<std::string, bool> map_bool;
+static map_string kernel_groups;
+static map_string kernel_strings;
+static map_kernel kernels;
+static map_bool compiled_kernels;
+static std::string dummy_kernel = "__kernel void dummy() { };";
+
+/**
+ * Initalizes the global std::map variables that
+ * hold the OpenCL kernel sources, the groups to
+ * which each kernel is assigned to and the
+ * information about which kernel was already compiled.
+ *
+ */
+inline void init_kernel_groups() {
+  // To identify the kernel group
+  kernel_groups["transpose"] = "basic_matrix";
+  kernel_groups["copy"] = "basic_matrix";
+  kernel_groups["zeros"] = "basic_matrix";
+  kernel_groups["identity"] = "basic_matrix";
+  kernel_groups["copy_triangular"] = "basic_matrix";
+  kernel_groups["copy_triangular_transposed"] = "basic_matrix";
+  kernel_groups["add"] = "basic_matrix";
+  kernel_groups["subtract"] = "basic_matrix";
+  kernel_groups["copy_submatrix"] = "basic_matrix";
+  kernel_groups["scalar_mul_diagonal"] = "matrix_multiply";
+  kernel_groups["scalar_mul"] = "matrix_multiply";
+  kernel_groups["basic_multiply"] = "matrix_multiply";
+  kernel_groups["multiply_self_transposed"] = "matrix_multiply";
+  kernel_groups["lower_tri_inv_step1"] = "matrix_inverse";
+  kernel_groups["lower_tri_inv_step2"] = "matrix_inverse";
+  kernel_groups["lower_tri_inv_step3"] = "matrix_inverse";
+  kernel_groups["cholesky_block"] = "cholesky_decomposition";
+  kernel_groups["check_nan"] = "check_gpu";
+  kernel_groups["check_symmetric"] = "check_gpu";
+  kernel_groups["check_diagonal_zeros"] = "check_gpu";
+
+  kernel_groups["dummy"] = "timing";
+
+  // Kernel group strings
+  // the dummy kernel is the only one not included in files
+  // so it is treated before the loop that iterates
+  // through  kernels to load all
+
+  kernel_strings["timing"] = dummy_kernel;
+  kernel_strings["check_gpu"] =
+#include <stan/math/prim/mat/kern_gpu/check_gpu.cl>  // NOLINT
```

Comment:
Rather than creating a folder at the `stan/math/prim/mat` level, I'd prefer to create one at the `stan/math/prim/mat/fun` level because as far as I can tell, these are only used here? We can look at how the compiler is arranged.

Reply:
done